### PR TITLE
Add new rules to PMD - design part 4

### PR DESCRIPTION
**Pull request by kanghj**
Omited UseVarargs for now, as it looks like quite a lot of effort for very little gain. 

Most of our *Attributes classes violates UncommentedEmptyConstructor, but I'm not sure how to deal with it correctly, as the no args constructor is used in many places.

### CODE REVIEW

File: src/main/java/teammates/common/util/ActivityLogEntry.java [hunk 1]
```diff
@@ -422,9 +427,11 @@ public String getTableCellColorCode(Long timeTaken) {
     public String getLogEntryActionsButtonClass() {
         
         String className = "";
-        if (message.toLowerCase().contains(Const.ACTION_RESULT_FAILURE.toLowerCase())) {
+        if (message.toLowerCase(Locale.ENGLISH).contains(
+                                        Const.ACTION_RESULT_FAILURE.toLowerCase(Locale.ENGLISH))) {
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/util/ActivityLogEntry.java]: Maybe 
```
message.toLowerCase(Locale.ENGLISH)
       .contains(Const.ACTION_RESULT_FAILURE.toLowerCase(Locale.ENGLISH))) {
```
instead:)?

File: src/main/java/teammates/common/util/ActivityLogEntry.java [hunk 2]
```diff
@@ -363,8 +366,9 @@ public String getPersonInfo() {
     public String getActionInfo() {
         String style = "";
         
-        if (message.toLowerCase().contains(Const.ACTION_RESULT_FAILURE.toLowerCase())
-           || message.toLowerCase().contains(Const.ACTION_RESULT_SYSTEM_ERROR_REPORT.toLowerCase())) {
+        if (message.toLowerCase(Locale.ENGLISH).contains(Const.ACTION_RESULT_FAILURE.toLowerCase(Locale.ENGLISH))
+           || message.toLowerCase(Locale.ENGLISH).contains(
+                                           Const.ACTION_RESULT_SYSTEM_ERROR_REPORT.toLowerCase(Locale.ENGLISH))) {
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/util/ActivityLogEntry.java]:
```
message.toLowerCase(Locale.ENGLISH)
       .contains(Const.ACTION_RESULT_SYSTEM_ERROR_REPORT.toLowerCase(Locale.ENGLISH))) {
```
perhaps

File: src/main/java/teammates/common/util/ActivityLogEntry.java [hunk 3]
```diff
@@ -375,10 +379,11 @@ public String getActionInfo() {
     
     public String getMessageInfo() {
         
-        if (message.toLowerCase().contains(Const.ACTION_RESULT_FAILURE.toLowerCase())) {
+        if (message.toLowerCase(Locale.ENGLISH).contains(Const.ACTION_RESULT_FAILURE.toLowerCase(Locale.ENGLISH))) {
             message = message.replace(Const.ACTION_RESULT_FAILURE, "<span class=\"text-danger\"><strong>" + Const.ACTION_RESULT_FAILURE + "</strong><br>");
             message = message + "</span><br>";
-        } else if (message.toLowerCase().contains(Const.ACTION_RESULT_SYSTEM_ERROR_REPORT.toLowerCase())) {
+        } else if (message.toLowerCase(Locale.ENGLISH).contains(
+                                        Const.ACTION_RESULT_SYSTEM_ERROR_REPORT.toLowerCase(Locale.ENGLISH))) {
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/util/ActivityLogEntry.java]: Could align the `toLowerCase` with `contains` here too

File: src/main/java/teammates/common/util/ActivityLogEntry.java [hunk 4]
```diff
@@ -422,9 +427,11 @@ public String getTableCellColorCode(Long timeTaken) {
     public String getLogEntryActionsButtonClass() {
         
         String className = "";
-        if (message.toLowerCase().contains(Const.ACTION_RESULT_FAILURE.toLowerCase())) {
+        if (message.toLowerCase(Locale.ENGLISH).contains(
+                                        Const.ACTION_RESULT_FAILURE.toLowerCase(Locale.ENGLISH))) {
             className = "btn-warning";
-        } else if (message.toLowerCase().contains(Const.ACTION_RESULT_SYSTEM_ERROR_REPORT.toLowerCase())) {
+        } else if (message.toLowerCase(Locale.ENGLISH).contains(
+                                        Const.ACTION_RESULT_SYSTEM_ERROR_REPORT.toLowerCase(Locale.ENGLISH))) {
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/util/ActivityLogEntry.java]: Here too

File: src/main/java/teammates/common/util/EmailLogEntry.java [hunk 5]
```diff
@@ -128,9 +129,11 @@ private String hightlightTextWithKeyWords(String[] keyStringsToHighlight, String
         }
         
         for (String stringToHighlight : keyStringsToHighlight) {
-            if (text.toLowerCase().contains(stringToHighlight.toLowerCase())) {
+            if (text.toLowerCase(Locale.ENGLISH).contains(
+                                            stringToHighlight.toLowerCase(Locale.ENGLISH))) {
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/util/EmailLogEntry.java]: Here too

File: src/main/java/teammates/common/util/EmailLogEntry.java [hunk 6]
```diff
@@ -128,9 +129,11 @@ private String hightlightTextWithKeyWords(String[] keyStringsToHighlight, String
...
                 
-                int startIndex = text.toLowerCase().indexOf(stringToHighlight.toLowerCase());
+                int startIndex = text.toLowerCase(Locale.ENGLISH).indexOf(
+                                                                    stringToHighlight.toLowerCase(Locale.ENGLISH));
```
Comment (JosephineKwa) [inline src/main/java/teammates/common/util/EmailLogEntry.java]: Here too

File: src/main/java/teammates/ui/controller/AdminActivityLogPageData.java [hunk 7]
```diff
@@ -223,7 +227,8 @@ public ActivityLogEntry filterLogs(ActivityLogEntry logEntry) {
         if (q.isInfoInQuery) {
             
             for (String keyString : q.infoValues) {
-                if (!logEntry.getMessageInfo().toLowerCase().contains(keyString.toLowerCase())) {
+                if (!logEntry.getMessageInfo().toLowerCase(Locale.ENGLISH).contains(
+                                                                               keyString.toLowerCase(Locale.ENGLISH))) {
```
Comment (JosephineKwa) [inline src/main/java/teammates/ui/controller/AdminActivityLogPageData.java]: Here as well

**Comment by wkurniawan07** [comment]
Chipping in a bit: similar to the other PR, is `UseLocaleWithCaseConversions` worth the trouble?

**Comment by kanghj** [comment]
Since we removed Locale in the other PR, I removed the rule here too.

**Comment by JosephineKwa** [comment]
Looks good!

**Pull request merged**